src/ethereum/istanbul/vm/instructions/environment.py

```diff
@@ -0,0 +1,550 @@
+
+    value = code[code_start_index : code_start_index + size]
+    # But it is possible that code_start_index + size won't exist in evm.code
+    # in which case we need to right pad the above obtained bytes with 0.
+    value = value.ljust(size, b"\x00")
+
+    memory_write(evm.memory, memory_start_index, value)
+
+
+def returndatasize(evm: Evm) -> None:
+    """
+    Pushes the size of the return data buffer onto the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_BASE)
+    return_size = U256(len(evm.return_data))
+    push(evm.stack, return_size)
+    evm.pc += 1
+
+
+def returndatacopy(evm: Evm) -> None:
+    """
+    Copies data from the return data buffer code to memory
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    memory_start_index = Uint(pop(evm.stack))
+    return_data_start_position = Uint(pop(evm.stack))
+    size = pop(evm.stack)
+    ensure(
+        return_data_start_position + size <= len(evm.return_data),
+        OutOfBoundsRead,
+    )
+
+    words = ceil32(Uint(size)) // 32
+    copy_gas_cost = u256_safe_multiply(
+        GAS_RETURN_DATA_COPY,
+        words,
+        exception_type=OutOfGasError,
+    )
+    memory_extend_gas_cost = calculate_gas_extend_memory(
+        evm.memory, memory_start_index, size
+    )
+    total_gas_cost = u256_safe_add(
+        GAS_VERY_LOW,
+        copy_gas_cost,
+        memory_extend_gas_cost,
+        exception_type=OutOfGasError,
+    )
+    evm.gas_left = subtract_gas(evm.gas_left, total_gas_cost)
+
+    extend_memory(evm.memory, memory_start_index, size)
+    value = evm.return_data[
+        return_data_start_position : return_data_start_position + size
+    ]
+    memory_write(evm.memory, memory_start_index, value)
+    evm.pc += 1
+
+
+def extcodehash(evm: Evm) -> None:
+    """
+    Returns the keccak256 hash of a contract’s bytecode
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    address = to_address(pop(evm.stack))
+
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_CODE_HASH)
+
+    evm.pc += 1
+
+    account = get_account(evm.env.state, address)
+
+    if account == EMPTY_ACCOUNT:
+        push(evm.stack, U256(0))
+    else:
+        code = account.code
+        code_hash = keccak256(code)
+        push(evm.stack, U256.from_be_bytes(code_hash))
+
+
+def self_balance(evm: Evm) -> None:
+    """
+    Pushes the balance of the current address to the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    :py:class:`~ethereum.istanbul.vm.error.OutOfGasError`
+        If `evm.gas_left` is less than GAS_FAST_STEP.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_FAST_STEP)
+
+    current_target = U256.from_be_bytes(evm.message.current_target)
+
+    address = to_address(current_target)
```

Comment:
Isn't `evm.message.current_target` an `Address` already? Why the round trip through `U256`?

Reply:
Ah, yes. You are right. I'll update this.

Reply:
Updated